docs/CodeTransformation.md

```diff
@@ -0,0 +1,94 @@
+
+Jest supports this via the [`transform` configuration option](Configuration.md#transform-objectstring-pathtotransformer--pathtotransformer-object).
+
+A transformer is a module that provides a synchronous function for transforming source files. For example, if you wanted to be able to use a new language feature in your modules or tests that aren't yet supported by Node, you might plug in one of many compilers that compile a future version of JavaScript to a current one.
+
+Jest will cache the result of a transformation and attempt to invalidate that result based on a number of factors, such as the source of the file being transformed and changing configuration.
+
+## Defaults
+
+Jest ships with one transformer out of the box - `babel-jest`. It will automatically load your project's Babel configuration and transform any file matching the following RegEx: `/\.[jt]sx?$/` meaning any `.js`, `.jsx`, `.ts` and `.tsx` file. In addition, `babel-jest` will inject the Babel plugin necessary for mock hoisting talked about in [ES Module mocking](ManualMocks.md#using-with-es-module-imports).
+
+If you override the `transform` configuration option `babel-jest` will no longer be active, and you'll need to add it manually if you wish to use Babel.
+
+## Writing custom transformers
+
+You can write you own transformer. The API of a transformer is as follows:
+
+```ts
+interface Transformer {
+  canInstrument?: boolean;
+  createTransformer?: (options?: unknown) => Transformer;
+
+  getCacheKey?: (
+    sourceText: string,
+    sourcePath: string,
+    options: TransformOptions,
+  ) => string;
+
+  process: (
+    sourceText: string,
+    sourcePath: string,
+    options: TransformOptions,
+  ) => TransformedSource;
+}
+
+interface TransformOptions {
+  config: Config.ProjectConfig;
+  /** A stringified version of the configuration - useful in cache busting */
+  configString: string;
+  instrument: boolean;
+  // names are copied from babel: https://babeljs.io/docs/en/options#caller
+  supportsDynamicImport: boolean;
+  supportsExportNamespaceFrom: boolean;
+  supportsStaticESM: boolean;
+  supportsTopLevelAwait: boolean;
+}
+
+type TransformedSource =
+  | {code: string; map?: RawSourceMap | string | null}
+  | string;
+
+// Config.ProjectConfig can be seen in in code [here](https://github.com/facebook/jest/blob/v26.6.3/packages/jest-types/src/Config.ts#L323)
+// RawSourceMap comes from [`source-map`](https://github.com/mozilla/source-map/blob/0.6.1/source-map.d.ts#L6-L12)
+```
+
+As can be seen, only `process` is mandatory to implement, although we highly recommend implementing `getCacheKey` as well, so we don't waste resources transpiling the same source file when we can read its previous result from disk.
```

Comment:
Point to the cache key package we recently added

Reply:
[`26e213f` (#10834)](https://github.com/facebook/jest/pull/10834/commits/26e213f20aaa21fa9df7666c1ea9789d3bd02591)